replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/HeartbeaterAndVerifierImpl.java

```diff
@@ -73,23 +135,307 @@ public void cleanUp() {
 
   @Override
   public void verify(ReplicationSite site) {
-    final String url = site.getUrl();
+    if (site.getVerifiedUrl() == null) {
+      final String url = site.getUrl();
+      HeartbeatResult heartbeatResult = doDiscovery(url);
 
-    // For now, fake the verification until implemented
-    site.setVerifiedUrl(url);
-    siteManager.save(site);
+      if (heartbeatResult.isHeartbeatSupported()) {
+        site.setRemoteManaged(true);
+        site.setVerifiedUrl(url);
+        siteManager.save(site);
+      } else if (heartbeatResult.isTmpRedirect()) {
+        HeartbeatResult redirectResult = doDiscovery(heartbeatResult.getRedirectUrl());
+        if (redirectResult.isHeartbeatSupported()) {
+          site.setRemoteManaged(true);
+          site.setVerifiedUrl(url);
+          siteManager.save(site);
+        }
+        // do nothing and retry the heartbeat next time
+      } else if (heartbeatResult.isPermRedirect()) {
```

Comment:
Why do we have a perm and tmp redirect. Do we care? Doesn't look like we will attempt to reverify either way.

Reply:
If it is a permanent or temporary redirect, a heartbeat is sent again to the redirect URL. If we hit a proxy we will need to follow the redirect.

Reply:
Correct. All redirects must be followed by designed (HTTP spec) up to a max number of redirects. We care about permanent a bit more as we need to save the new location for subsequent messages.